xen/arch/arm/cpufreq/scpi_cpufreq.c

```diff
@@ -0,0 +1,353 @@
+    int i;
+
+    for ( i = 0; i < 100; i++ )
+    {
+        curr_freq = scpi_cpufreq_get(cpu);
+        if ( curr_freq == freq )
+            return 1;
+        udelay(10);
+    }
+    return 0;
+}
+
+static int scpi_cpufreq_target(struct cpufreq_policy *policy,
+                               unsigned int target_freq, unsigned int relation)
+{
+    struct scpi_cpufreq_data *data = cpufreq_driver_data[policy->cpu];
+    struct processor_performance *perf;
+    struct cpufreq_freqs freqs;
+    cpumask_t online_policy_cpus;
+    unsigned int next_state = 0; /* Index into freq_table */
+    unsigned int next_perf_state = 0; /* Index into perf table */
+    unsigned int j;
+    int result = 0;
+    const struct scpi_opp *opp;
+    int idx, max_opp;
+
+    if ( unlikely(!data) || !data->perf || !data->freq_table )
+        return -ENODEV;
+
+    perf = data->perf;
+    result = cpufreq_frequency_table_target(policy,
+                                            data->freq_table,
+                                            target_freq,
+                                            relation, &next_state);
+    if ( unlikely(result) )
+        return -ENODEV;
+
+    cpumask_and(&online_policy_cpus, &cpu_online_map, policy->cpus);
+
+    next_perf_state = data->freq_table[next_state].index;
+    if ( perf->state == next_perf_state )
+    {
+        if ( unlikely(policy->resume) )
+            policy->resume = 0;
+        else
+            return 0;
+    }
+
+    freqs.old = perf->states[perf->state].core_frequency * 1000;
+    freqs.new = data->freq_table[next_state].frequency;
+
+    /* Sanity check. TODO Put this at the beginning. */
+    if ( data->domain_id < 0 || !data->info )
+        return -EINVAL;
+
+    max_opp = data->info->count;
+    opp = data->info->opps;
+
+    for ( idx = 0; idx < max_opp; idx++, opp++ )
+    {
+        if ( opp->freq == freqs.new * 1000 /* Hz */ )
+            break;
+    }
+
+    if ( idx == max_opp )
+        return -EINVAL;
+
+    result = scpi_ops->dvfs_set_idx(data->domain_id, (u8)idx);
+    if ( result < 0 )
+        return result;
+
+    /* TODO x86 checks acpi_pstate_strict here */
+    if ( 1 && !check_freq(policy->cpu, freqs.new) )
+    {
+        printk(KERN_WARNING "Fail transfer to new freq %d\n", freqs.new);
+        return -EAGAIN;
+    }
+
+    for_each_cpu( j, &online_policy_cpus )
+        cpufreq_statistic_update(j, perf->state, next_perf_state);
+
+    perf->state = next_perf_state;
+    policy->cur = freqs.new;
+
+    return result;
+}
+
+static int scpi_cpufreq_verify(struct cpufreq_policy *policy)
+{
+    struct scpi_cpufreq_data *data;
+    struct processor_performance *perf;
+
+    if ( !policy || !(data = cpufreq_driver_data[policy->cpu] ) ||
+         !processor_pminfo[policy->cpu] )
+        return -EINVAL;
+
+    perf = &processor_pminfo[policy->cpu]->perf;
+
+    cpufreq_verify_within_limits(policy, 0,
+        perf->states[perf->platform_limit].core_frequency * 1000);
```

Comment:
Why you constantly converting Hz to kHz? Can't you store all frequencies in one unit?

Reply:
Let me elaborate a bit more about this:

1. Frequency table keeps frequency in kHz
See cpufreq.h:
struct cpufreq_frequency_table {
    unsigned int    index;     /* any */
    unsigned int    frequency; /* kHz - doesn't need to be in ascending
                                * order */
};
2. Performance states keep frequency in MHz
See public/platform.h:
struct xen_processor_px {
    uint64_t core_frequency; /* megahertz */
    uint64_t power;      /* milliWatts */
    uint64_t transition_latency; /* microseconds */
    uint64_t bus_master_latency; /* microseconds */
    uint64_t control;        /* control value */
    uint64_t status;     /* success indicator */
};
3. SCPI protocol defines frequencies to be in Hz:
See "3.2.9 Get DVFS Info" for example:
http://infocenter.arm.com/help/topic/com.arm.doc.dui0922g/scp_message_interface_v1_2_DUI0922G_en.pdf 
Frequency in Hertz (Hz)

Reply:
Yeah. But you can select your units to minimize number of conversions. Looks like cpufreq in xen uses kHz. So, you can make your public SCPI API to use kHz and then convert to Hz directly before calling SCPI.
ATM it is very confusing and it is very easy to forget about units somewhere. 
Have you heard story of Mars Climate Orbiter?

Reply:
No, I haven't)

Reply:
Software that calculated the total impulse produced by thruster firings calculated results in pound-seconds. The trajectory calculation software then used these results - expected to be in newton-seconds - to update the predicted position of the spacecraft. Spacecraft crashed into Mars.

Reply:
then probably we can(?) add suffixes like bla_bla_khz, _hz, _mhz?

Reply:
I would avoid any changes to common Xen headers for now. i.e. public headers like platform.h. I will add comments for each conversions.